mopidy/local/playlists.py

```diff
@@ -16,70 +15,62 @@
 logger = logging.getLogger(__name__)
 
 
+def _ref(playlist):
+    return Ref.playlist(uri=playlist.uri, name=playlist.name)
+
+
 class LocalPlaylistsProvider(backend.PlaylistsProvider):
     def __init__(self, *args, **kwargs):
         super(LocalPlaylistsProvider, self).__init__(*args, **kwargs)
         self._media_dir = self.backend.config['local']['media_dir']
         self._playlists_dir = self.backend.config['local']['playlists_dir']
-        self._playlists = []
+        self._playlists = {}
         self.refresh()
 
+    def get_playlists(self, ref=True):
+        if ref:
+            playlists = map(_ref, self._playlists.values())
+        else:
+            playlists = self._playlists.values()
+        return sorted(playlists, key=operator.attrgetter('name'))
+
     @property
     def playlists(self):
-        return copy.copy(self._playlists)
+        return self.get_playlists(ref=False)
 
     @playlists.setter
     def playlists(self, playlists):
-        self._playlists = playlists
+        self._playlists = {playlist.uri: playlist for playlist in playlists}
 
     def create(self, name):
         playlist = self._save_m3u(Playlist(name=name))
-        old_playlist = self.lookup(playlist.uri)
-        if old_playlist is not None:
-            index = self._playlists.index(old_playlist)
-            self._playlists[index] = playlist
-        else:
-            self._playlists.append(playlist)
-        self._playlists.sort(key=operator.attrgetter('name'))
-        logger.info('Created playlist %s', playlist.uri)
+        self._playlists[playlist.uri] = playlist
         return playlist
 
     def delete(self, uri):
-        playlist = self.lookup(uri)
-        if not playlist:
-            logger.warn('Trying to delete unknown playlist %s', uri)
-            return
-        path = local_playlist_uri_to_path(uri, self._playlists_dir)
-        if os.path.exists(path):
-            os.remove(path)
+        if uri in self._playlists:
+            path = local_playlist_uri_to_path(uri, self._playlists_dir)
+            if os.path.exists(path):
+                os.remove(path)
+            else:
+                logger.warn('Trying to delete missing playlist file %s', path)
+            del self._playlists[uri]
         else:
-            logger.warn('Trying to delete missing playlist file %s', path)
-        self._playlists.remove(playlist)
+            logger.warn('Trying to delete unknown playlist %s', uri)
 
     def lookup(self, uri):
-        # TODO: store as {uri: playlist} when get_playlists() gets
-        # implemented
-        for playlist in self._playlists:
-            if playlist.uri == uri:
-                return playlist
+        return self._playlists[uri]
```

Comment:
This would crash on an unknown URI. Using `.get(uri)`, and it follow the contract of returning `None` if not found.